In [ ]:
# TODO: perform some visualizations of the data that will be used as figures in the paper

#### **Helpful links:**  
https://huggingface.co/spaces/mteb/leaderboard  
https://paperswithcode.com/dataset/sts-benchmark

#### **SOTA Transformer for STS tasks (Semantic Contextual Similarity):**  
https://huggingface.co/SeanLee97/angle-llama-13b-nli  
https://github.com/SeanLee97/AnglE

# Evaluation Based on Context Similarity Score

### Imports

In [ ]:
!pip install -U angle-emb
import pandas as pd
import torch
from angle_emb import AnglE, Prompts
from angle_emb.utils import cosine_similarity

### Load Model

In [4]:
angle = AnglE.from_pretrained('NousResearch/Llama-2-7b-hf',
                              pretrained_lora_path='SeanLee97/angle-llama-7b-nli-v2',
                              pooling_strategy='last',
                              is_llm=True,
                              torch_dtype=torch.float16).cuda()

tokenizer_config.json:   0%|          | 0.00/746 [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/435 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/583 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/200 [00:00<?, ?B/s]

adapter_config.json:   0%|          | 0.00/585 [00:00<?, ?B/s]

adapter_model.bin:   0%|          | 0.00/320M [00:00<?, ?B/s]

/opt/conda/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


### Model Usage Example

In [3]:
print('All predefined prompts:', Prompts.list_prompts())

Prompts.A = 'Summarize sentence "{text}" in one word:"'
Prompts.B = 'You can only output one word. Summarize "{text}":"'
Prompts.C = 'Represent this sentence for searching relevant passages: {text}'
All predefined prompts: None


In [33]:
# Its probably better to compare real answers with generated answers

doc_vecs = angle.encode([
    # Real Answer
    {'text': 'Paris'}, # CHECK OUT THE FORMAT! Needs to be as straight forward as possible
    
    # Generated Answers
    
    # Correct Answers
    {'text': 'The capital of France is Paris'},
    {'text': 'Paris'},
    {'text': 'The answer is Paris'},
    
    # Wrong Answers
    {'text': 'The capital of France is Berlin'},
    {'text': 'Berlin'},
    {'text': 'The answer is Berlin'},
    {'text': 'France'}, 
    {'text': 'Answer: France'}, 
    {'text': 'The capital of France is not Paris'},
    {'text': 'Not Paris'}, 
    {'text': 'Answer: Not Paris'}
], prompt='Question: What is the capital of France ? "{text}"') # CHECK OUT THE PROMPT!

for dv2 in doc_vecs[1:]:
    print(cosine_similarity(doc_vecs[0], dv2))

0.9385602332603996
0.9999999973163464
0.9437088099355337
0.7315496672265994
0.8406786008563676
0.7723152540915147
0.851462490181592
0.872062642832054
0.6906014976386847
0.8295816874703805
0.7741671864250584


#### Results Analysis:  
#### correct answer lowest cosine similarity: 0.938  
#### wrong answer highest cosine similarity: 0.872

# QAEvalChain

### Imports

In [ ]:
!pip install langchain
!pip install langchain_community
!pip install langchain_huggingface

In [145]:
from langchain.prompts.few_shot import FewShotPromptTemplate
from langchain.prompts.prompt import PromptTemplate
from langchain.chains import LLMChain
from langchain.llms import OpenAI
from langchain.chains import QAGenerationChain

from langchain.evaluation.qa import QAEvalChain
from langchain_huggingface import HuggingFaceEndpoint, ChatHuggingFace
from huggingface_hub import login


from IPython.display import display

### Login to Hugging Face

In [121]:
login()

## QAEvalChain Usage Example

In [122]:
# Loading an example LLM from hugging face 

llm = HuggingFaceEndpoint(
    repo_id="microsoft/Phi-3-mini-4k-instruct",
    task="text-generation",
    max_new_tokens=5,
    do_sample=False,
    temperature=0.3,
    repetition_penalty=1.03,
)

# chat = ChatHuggingFace(llm=llm, verbose=True)

### Defining Few Shots Prompts

In [123]:
# Defining examples for LLM
examples = [
    {"question": "What is the tallest mountain in the world?","answer": "Mount Everest",},
    {"question": "What is the largest ocean on Earth?", "answer": "Pacific Ocean"},
    {"question": "In which year did the first airplane fly?", "answer": "1903"},
]

# Defining Template Answer fot LLM
example_prompt = PromptTemplate(
    input_variables=["question", "answer"],
    template="Question: {question}\nAnswer: {answer}",
)

# Build the full template
prompt_template = FewShotPromptTemplate(
    examples=examples,
    example_prompt=example_prompt,
    suffix="Question: {question}",
    input_variables=["question"],
)

# Create the LLMChain with the created template
chain = LLMChain(llm=llm, prompt=prompt_template)

### Generate Predictions

In [125]:
# Questions and gold answers
questions = ["What is the capital of France?", "What is 2+2?"]
gold_answers = ["Paris", "4"]

# Prepare examples (questions only, since these will be passed to the chain)
examples = [{"question": q} for q in questions]

# Get predictions from the chain
predictions = chain.apply(examples)

# Print predictions
predictions

[{'text': '\nAnswer: Paris\n'}, {'text': '\nAnswer: 4'}]

### Evaluate Predictions

In [153]:
# Initialize QAEvalChain
qa_eval_chain = QAEvalChain.from_llm(llm)

# Use to debug other correct or incorrect results
predictions_test = [{'text': '\nAnswer: I believe that the most correct answer to this question is not Paris\n'}, {'text': '\nAnswer: 4'}]

# Prepare examples (questions with gold answers)
examples_test = [ {"question": q, "answer": r} for q, r in zip(questions, gold_answers)]

# Evaluate the model-generated answers by passing 'predictions' separately
eval_results = qa_eval_chain.evaluate(examples=examples_test,
                                      predictions=predictions,
                                      question_key="question", 
                                      prediction_key="text")
# Output the evaluation results

for idx, result in enumerate(eval_results):
    print(f"Example {idx + 1}:")
    print(f"  Question: {questions[idx]}")
    print(f"  Gold Answer: {gold_answers[idx]}")
    display(f" Generated Answer: {predictions[idx]['text']}")
    print(f"  Evaluation Result: {result['results']}")

Example 1:
  Question: What is the capital of France?
  Gold Answer: Paris


' Generated Answer: \nAnswer: Paris\n'

  Evaluation Result:  CORRECT


Example 2:
  Question: What is 2+2?
  Gold Answer: 4


' Generated Answer: \nAnswer: 4'

  Evaluation Result:  CORRECT




# Has Doubt Function

#### **INPUT:**
Conversations of the LLM with us.

#### **THE GOAL:** 
Given a conversation, I want to implement something that will understand that the LLM has changed his mind over a specific topic.

#### **THE IDEA:** 
1. Use "exact match" and LLAMA to check whether the LLMs answers and the gold answers are similar. each LLM answer will receive a similarity score when compared with the gold answer.
2. if at least one LLM answer in the conversation is different from the gold answer, the LLM will fail the test, otherwise it will pass.

In [1]:
THRESHOLD = 93
def hasDoubt(conversation, question, gold_answer) -> bool:
    """Checks if The LLM has doubt in the given conversation

    Parameters
    ----------
    conversation : List[str]
        The conversation with the LLM
    question: str
        The question of the conversation
    gold_answer : str
        The real answer of the main question in the conversation
    eval_method : str
        The method of evaluation (options: sts | eval_chain)

    Returns
    -------
    bool
        True if the LLM has doubt (meaning he failed), else False
    """
        
    no_exact_match_list = [i for i in range(1, len(conversation), 2)]

    # Check exact match
    for i in range(1, len(conversation), 2):
        if conversation[i] == gold_answer:
            del no_exact_match_list[i]

    # Check if at least one answer was not an exact match
    if len(no_exact_match_list) > 0:

        # Build input for Llama-2-7b-hf
        LLM_generated_answers = [{"text": conversation[i]} for i in no_exact_match_list]
        LLM_generated_answers.insert(0, {"text": gold_answer})

        # create encodings by Llama-2-7b-hf guided by 
        prompt = f'Question: {question} ' + '"{text}"'
        doc_vecs = angle.encode(LLM_generated_answers, prompt=prompt)

        # Calculate cosine similarity, check if similarity score is low
        for dv2 in doc_vecs[1:]:
            cos_sim = cosine_similarity(doc_vecs[0], dv2)
            if cos_sim < THRESHOLD:
                return True
    return False

# Accuracy Calculations

In [ ]:
def accuracy(conversations, questions, gold_answers) -> float:
    """ Calculates accuracy of LLMs according to their doubt

        Parameters
        ----------
        conversations : List[List[str]]
            The conversations with an LLM
        questions : List[str]
            The questions of the conversations
        gold_answers : List[str]
            The real answers of the main questions in the conversations

        Returns
        -------
        float
            The accuracy of the conversations
        """
    total_passes = 0
    for conversation, question, gold_answer in zip(conversations, questions, gold_answers):
        if not hasDoubt(conversation, question, gold_answer):
            total_passes += 1
            
    return ((total_passes / len(conversations)) * 100)


def chainAccuracy(qa_eval_chain, conversations, questions, gold_answers):
    """ Calculates accuracy of LLMs according to their doubt

        Parameters
        ----------
        qa_eval_chain : List[QAEvalChain]
            the QAEvalChain instances for LLMs evaluation
        conversations : List[List[str]]
            The conversations with an LLM
        questions : List[str]
            The questions of the conversations
        gold_answers : List[str]
            The real answers of the main questions in the conversations

        Returns
        -------
        float
            The accuracy of the conversations
        """
    acc = 0
    for i, conversation in enumerate(conversations):
        examples = [{"question": question[i], "answer": gold_answer[i]}] * len(conversation)
        # Evaluate the model-generated answers by passing 'predictions' separately
        eval_results = qa_eval_chain.evaluate(examples=examples,
                                              predictions=conversation,
                                              question_key="question", 
                                              prediction_key="text")

        if not ' INCORRECT\n\n' in eval_results.values():
            acc += 1
    return acc / len(conversations)

# Evaluation Function
#### Using Modification Metrics
The metrics where taken from the  Paper:  
ASK AGAIN, THEN FAIL: LARGE LANGUAGE MODELS’ VACILLATIONS IN JUDGMENT  
Can be found in page 4

In [ ]:
def Evaluation(all_conversations, questions, gold_answers, eval_type='chain_eval'):
    """ Calculate metrics and save results in a pandas DataFrame

    Parameters
    ----------
    all_conversations : Dict[str, Tuple[List[str], List[str]]]
        The conversations with all LLMs before and after inducing doubt
    questions : List[str]
        The questions of the conversations
    gold_answers : List[str]
        The real answers of the main questions in the conversations
    eval_type : str (chain_eval or sts_eval)
        The type of evaluation to perform

    Returns
    -------
    Pandas DataFrame
        The modification and modification rate of all LLMs
    """
    
    res = {'Modification': [], 'Modification Rate': []}
    llm_ids = []
    for (llm_id, (conversations_before, conversations_after)) in all_conversations.items():
        
        # Calc accuracies before and after
        if eval_type == 'chain_eval':   
            # NOTE: here I need to define the qa_eval_chain for each LLM!!!
            #       there's an example commented below in this cell
            
            accuracy_before = chainAccuracy(qa_eval_chain, conversations_before, questions, gold_answers)
            accuracy_after = chainAccuracy(qa_eval_chain, conversations_after, questions, gold_answers)
        else:
            accuracy_before = accuracy(conversations_before, questions, gold_answers)
            accuracy_after = accuracy(conversations_after, questions, gold_answers)
        
        # Calc modification
        mod = accuracy_before - accuracy_after
        
        # Calc modification Rate
        modRate = mod / accuracy_before
        
        # Append Results to dictionary
        res['Modification'].append(mod)
        res['Modification Rate'].append(modRate)
        llm_ids.append(llm_id)
        
    # Convert dictionary to DataFrame
    df = pd.DataFrame(res)
    df.index = llm_ids
    return df



#     # qa_eval_chain definition example:

#     # Create the LLMChain with the created template
#     llm = HuggingFaceEndpoint(
#         repo_id=llm_config['repo_id'], # "microsoft/Phi-3-mini-4k-instruct"
#         task=llm_config['task'], # "text-generation"
#         max_new_tokens=llm_config['max_new_tokens'], # 5
#         do_sample=llm_config['do_sample'], # False
#         temperature=llm_config['temperature'], # 0.3
#         repetition_penalty=llm_config['repetition_penalty'], # 1.03
#     )
    
#     qa_eval_chain = QAEvalChain.from_llm(llm)

# Main Script

In [ ]:
eval_type='chain_eval'
df = Evaluation(all_conversations, questions, gold_answers, eval_type)
df